File: riveredge-backend/src/apps/kuaizhizao/schemas/delivery_project.py
"""交付项目 Pydantic 模型"""

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# --- 人员（负责人 + 成员）---

class DeliveryMemberInput(BaseModel):
    user_id: int
    user_name: Optional[str] = None


class DeliveryMemberResponse(BaseModel):
    user_id: int
    user_name: str


# --- 流程模板 ---

class DeliveryProcessTemplateNodeTaskBase(BaseModel):
    task_key: str
    task_name: str
    sort_order: int = 0
    default_owner_role: Optional[str] = None
    planned_duration_days: int = 0


class DeliveryProcessTemplateNodeTaskCreate(DeliveryProcessTemplateNodeTaskBase):
    pass


class DeliveryProcessTemplateNodeTaskResponse(DeliveryProcessTemplateNodeTaskBase):
    id: int
    template_node_id: int

    class Config:
        from_attributes = True


class DeliveryProcessTemplateNodeBase(BaseModel):
    node_key: str
    node_name: str
    sort_order: int = 0
    default_owner_role: Optional[str] = None
    planned_duration_days: int = 0
    is_critical: bool = False
    is_milestone: bool = False
    tasks: List[DeliveryProcessTemplateNodeTaskCreate] = Field(default_factory=list)


class DeliveryProcessTemplateNodeCreate(DeliveryProcessTemplateNodeBase):
    pass


class DeliveryProcessTemplateNodeResponse(BaseModel):
    id: int
    template_id: int
    node_key: str
    node_name: str
    sort_order: int = 0
    default_owner_role: Optional[str] = None
    planned_duration_days: int = 0
    is_critical: bool = False
    is_milestone: bool = False
    tasks: List[DeliveryProcessTemplateNodeTaskResponse] = Field(default_factory=list)

    class Config:
        from_attributes = True


class DeliveryProcessTemplateCreate(BaseModel):
    template_name: str
    project_type: Optional[str] = None
    is_active: bool = True
    notes: Optional[str] = None
    nodes: List[DeliveryProcessTemplateNodeCreate] = Field(default_factory=list)


class DeliveryProcessTemplateUpdate(BaseModel):
    template_name: Optional[str] = None
    project_type: Optional[str] = None
    is_active: Optional[bool] = None
    notes: Optional[str] = None
    nodes: Optional[List[DeliveryProcessTemplateNodeCreate]] = None


class DeliveryProcessTemplateResponse(BaseModel):
    id: int
    template_code: str
    template_name: str
    project_type: Optional[str] = None
    is_active: bool
    is_default: bool
    notes: Optional[str] = None
    nodes: List[DeliveryProcessTemplateNodeResponse] = Field(default_factory=list)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class DeliveryProcessTemplateListEnvelope(BaseModel):
    items: List[DeliveryProcessTemplateResponse]
    total: int


# --- 交付项目 ---

class DeliveryProjectNodeTaskResponse(BaseModel):
    id: int
    project_id: int
    node_id: int
    template_task_id: Optional[int] = None
    task_key: Optional[str] = None
    task_name: str
    sort_order: int = 0
    status: str
    owner_id: Optional[int] = None
    owner_name: Optional[str] = None
    members: List[DeliveryMemberResponse] = Field(default_factory=list)
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None
    actual_start_date: Optional[date] = None
    actual_end_date: Optional[date] = None
    progress_percent: Decimal = Decimal("0")

    class Config:
        from_attributes = True


class DeliveryProjectNodeTaskCreate(BaseModel):
    node_id: int
    task_name: str
    sort_order: int = 0
    owner_id: Optional[int] = None
    members: List[DeliveryMemberInput] = Field(default_factory=list)
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None


class DeliveryProjectNodeTaskUpdate(BaseModel):
    task_name: Optional[str] = None
    sort_order: Optional[int] = None
    status: Optional[str] = None
    owner_id: Optional[int] = None
    members: Optional[List[DeliveryMemberInput]] = None
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None
    progress_percent: Optional[Decimal] = None


class DeliveryProjectNodeResponse(BaseModel):
    id: int
    project_id: int
    node_key: str
    node_name: str
    sort_order: int
    status: str
    progress_percent: Decimal
    owner_id: Optional[int] = None
    owner_name: Optional[str] = None
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None
    actual_start_date: Optional[date] = None
    actual_end_date: Optional[date] = None
    is_critical: bool
    is_milestone: bool
    tasks: List[DeliveryProjectNodeTaskResponse] = Field(default_factory=list)

    class Config:
        from_attributes = True


class DeliveryProjectCreate(BaseModel):
    project_name: str
    process_template_id: Optional[int] = None
    sales_order_id: Optional[int] = None
    customer_id: Optional[int] = None
    delivery_date: Optional[date] = None
    owner_id: Optional[int] = None
    members: List[DeliveryMemberInput] = Field(default_factory=list)
    material_id: Optional[int] = None
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    material_spec: Optional[str] = None
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None
    notes: Optional[str] = None


class DeliveryProjectUpdate(BaseModel):
    project_name: Optional[str] = None
    delivery_date: Optional[date] = None
    owner_id: Optional[int] = None
    members: Optional[List[DeliveryMemberInput]] = None
    status: Optional[str] = None
    notes: Optional[str] = None
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None


class DeliveryProjectCompleteRequest(BaseModel):
    force: bool = False
    reason: Optional[str] = None


class DeliveryProjectChangeTemplateRequest(BaseModel):
    process_template_id: int


class DeliveryProjectNodeUpdate(BaseModel):
    owner_id: Optional[int] = None


class DeliveryProjectResponse(BaseModel):
    id: int
    project_code: str
    project_name: str
    process_template_id: Optional[int] = None
    process_template_name: Optional[str] = None
    sales_order_id: Optional[int] = None
    sales_order_code: Optional[str] = None
    customer_id: Optional[int] = None
    customer_name: Optional[str] = None
    delivery_date: Optional[date] = None
    owner_id: Optional[int] = None
    owner_name: Optional[str] = None
    members: List[DeliveryMemberResponse] = Field(default_factory=list)
    material_id: Optional[int] = None
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    material_spec: Optional[str] = None
    material_lines: List[Dict[str, Any]] = Field(default_factory=list)
    rd_project_id: Optional[int] = None
    status: str
    progress_percent: Decimal
    current_node_key: Optional[str] = None
    current_node_name: Optional[str] = None
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None
    actual_start_date: Optional[date] = None
    actual_end_date: Optional[date] = None
    notes: Optional[str] = None
    nodes: List[DeliveryProjectNodeResponse] = Field(default_factory=list)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None

    class Config:
        from_attributes = True


class DeliveryProjectListResponse(BaseModel):
    id: int
    project_code: str
    project_name: str
    sales_order_code: Optional[str] = None
    customer_name: Optional[str] = None
    delivery_date: Optional[date] = None
    owner_name: Optional[str] = None
    member_count: int = 0
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    status: str
    progress_percent: Decimal
    current_node_name: Optional[str] = None
    nodes: List[DeliveryProjectNodeResponse] = Field(default_factory=list)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None

    class Config:
        from_attributes = True


class DeliveryLinkedRdProjectSummary(BaseModel):
    id: int
    project_code: str
    project_name: str


class DeliveryProjectListEnvelope(BaseModel):
    items: List[DeliveryProjectListResponse]
    total: int


class PushDeliveryProjectFromSalesOrderRequest(BaseModel):
    process_template_id: Optional[int] = None
    project_name: Optional[str] = None
    owner_id: Optional[int] = None


class PushDeliveryProjectPreviewResponse(BaseModel):
    sales_order_id: int
    sales_order_code: str
    customer_id: Optional[int] = None
    customer_name: Optional[str] = None
    delivery_date: Optional[date] = None
    material_lines: List[Dict[str, Any]] = Field(default_factory=list)
    existing_project_id: Optional[int] = None
    existing_project_code: Optional[str] = None
    default_template_id: Optional[int] = None
    default_template_name: Optional[str] = None


# --- 节点汇报 ---

class DeliveryNodeReportCreate(BaseModel):
    project_id: int
    node_id: int
    report_date: date
    progress_percent: Decimal = Field(default=Decimal("0"), ge=0, le=100)
    content: Optional[str] = None
    attachments: Optional[List[Dict[str, Any]]] = None


class DeliveryNodeReportUpdate(BaseModel):
    report_date: Optional[date] = None
    progress_percent: Optional[Decimal] = Field(default=None, ge=0, le=100)
    content: Optional[str] = None
    attachments: Optional[List[Dict[str, Any]]] = None


class DeliveryNodeReportResponse(BaseModel):
    id: int
    report_code: str
    project_id: int
    project_code: str
    node_id: int
    node_key: str
    node_name: str
    reporter_id: Optional[int] = None
    reporter_name: Optional[str] = None
    report_date: date
    progress_percent: Decimal
    content: Optional[str] = None
    attachments: Optional[List[Dict[str, Any]]] = None
    status: str
    reviewer_id: Optional[int] = None
    reviewer_name: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    review_notes: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None

    class Config:
        from_attributes = True


class DeliveryNodeReportListEnvelope(BaseModel):
    items: List[DeliveryNodeReportResponse]
    total: int


class DeliveryNodeReportReviewRequest(BaseModel):
    approved: bool
    review_notes: Optional[str] = None


# --- 问题跟踪 ---

class DeliveryIssueCreate(BaseModel):
    project_id: int
    node_id: Optional[int] = None
    issue_type: str = "other"
    priority: str = "normal"
    title: str
    description: Optional[str] = None
    assignee_id: Optional[int] = None
    due_date: Optional[date] = None


class DeliveryIssueUpdate(BaseModel):
    issue_type: Optional[str] = None
    priority: Optional[str] = None
    status: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    assignee_id: Optional[int] = None
    due_date: Optional[date] = None
    resolution: Optional[str] = None


class DeliveryIssueResponse(BaseModel):
    id: int
    issue_code: str
    project_id: int
    project_code: str
    node_id: Optional[int] = None
    node_name: Optional[str] = None
    issue_type: str
    priority: str
    status: str
    title: str
    description: Optional[str] = None
    assignee_id: Optional[int] = None
    assignee_name: Optional[str] = None
    due_date: Optional[date] = None
    resolved_at: Optional[datetime] = None
    resolution: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None

    class Config:
        from_attributes = True


class DeliveryIssueListEnvelope(BaseModel):
    items: List[DeliveryIssueResponse]
    total: int


class DeliveryProjectWorkbenchResponse(DeliveryProjectResponse):
    recent_reports: List[DeliveryNodeReportResponse] = Field(default_factory=list)
    open_issues: List[DeliveryIssueResponse] = Field(default_factory=list)
    linked_rd_project: Optional[DeliveryLinkedRdProjectSummary] = None


# --- 交付中心 / 跟进表 ---

class DeliveryDashboardKpi(BaseModel):
    active_projects: int = 0
    overdue_nodes: int = 0
    at_risk_projects: int = 0
    open_issues: int = 0


class DeliveryGanttItem(BaseModel):
    """交付中心甘特图行（项目节点）"""

    id: int = Field(description="甘特行唯一 ID（project_id * 100000 + node_id）")
    project_id: int
    node_id: int
    project_code: str
    project_name: str
    node_name: str
    customer_name: Optional[str] = None
    node_status: Optional[str] = None
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None
    progress: float = 0


class DeliveryDashboardResponse(BaseModel):
    kpis: DeliveryDashboardKpi
    recent_projects: List[DeliveryProjectListResponse] = Field(default_factory=list)
    overdue_nodes: List[Dict[str, Any]] = Field(default_factory=list)
    project_gantt: List[DeliveryGanttItem] = Field(default_factory=list)


class DeliveryFollowUpRow(BaseModel):
    project_id: int
    project_code: str
    project_name: str
    customer_name: Optional[str] = None
    delivery_date: Optional[date] = None
    status: str
    progress_percent: Decimal
    current_node_name: Optional[str] = None
    nodes: List[DeliveryProjectNodeResponse] = Field(default_factory=list)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None


class DeliveryFollowUpListEnvelope(BaseModel):
    items: List[DeliveryFollowUpRow]
    total: int


class DeliveryProgressSummaryRow(BaseModel):
    id: int
    project_code: str
    project_name: str
    customer_name: Optional[str] = None
    sales_order_code: Optional[str] = None
    delivery_date: Optional[date] = None
    owner_name: Optional[str] = None
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    status: str
    progress_percent: Decimal
    current_node_name: Optional[str] = None
    planned_end_date: Optional[date] = None
    overdue_node_count: int = 0
    open_issue_count: int = 0
    days_to_delivery: Optional[int] = None
    node_summary: Optional[str] = None

    class Config:
        from_attributes = True


class DeliveryProgressSummaryEnvelope(BaseModel):
    items: List[DeliveryProgressSummaryRow]
    total: int


class DeliveryProcessProgressRow(BaseModel):
    """项目流程进度表：一行一节点（含项目头信息）"""

    id: str
    project_id: int
    project_code: str
    project_name: str
    sales_order_code: Optional[str] = None
    customer_name: Optional[str] = None
    project_owner_name: Optional[str] = None
    material_name: Optional[str] = None
    delivery_date: Optional[date] = None
    node_id: int
    node_key: str
    node_name: str
    sort_order: int
    node_status: str
    progress_percent: Decimal
    node_owner_name: Optional[str] = None
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None
    actual_start_date: Optional[date] = None
    actual_end_date: Optional[date] = None
    reporter_name: Optional[str] = None
    issue_count: int = 0
    is_critical: bool = False
    is_milestone: bool = False


class DeliveryProcessProgressEnvelope(BaseModel):
    items: List[DeliveryProcessProgressRow]
    total: int


class DeliveryScheduleRow(BaseModel):
    """项目流程排单表"""

    project_id: int
    project_code: str
    project_name: str
    customer_name: Optional[str] = None
    delivery_date: Optional[date] = None
    owner_name: Optional[str] = None
    status: str
    progress_percent: Decimal
    current_node_name: Optional[str] = None
    schedule_node_name: Optional[str] = None
    schedule_node_owner_name: Optional[str] = None
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None
    node_status: Optional[str] = None
    report_overdue: bool = False
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None


class DeliveryScheduleListEnvelope(BaseModel):
    items: List[DeliveryScheduleRow]
    total: int


class DeliveryIssueProgressRow(BaseModel):
    """项目进度问题报表"""

    id: int
    issue_code: str
    project_code: str
    project_name: str
    customer_name: Optional[str] = None
    node_name: Optional[str] = None
    issue_type: str
    priority: str
    status: str
    title: str
    assignee_name: Optional[str] = None
    due_date: Optional[date] = None
    created_at: Optional[datetime] = None


class DeliveryIssueProgressEnvelope(BaseModel):
    items: List[DeliveryIssueProgressRow]
    total: int

File: riveredge-frontend/src/apps/kuaizhizao/services/delivery-project.ts
/**
 * 交付项目（订单交机）API
 */

import { apiRequest } from '../../../services/api';

export interface DeliveryMember {
  user_id: number;
  user_name: string;
}

export interface DeliveryProjectNodeTask {
  id: number;
  project_id: number;
  node_id: number;
  template_task_id?: number | null;
  task_key?: string | null;
  task_name: string;
  sort_order: number;
  status: string;
  owner_id?: number | null;
  owner_name?: string | null;
  members?: DeliveryMember[];
  planned_start_date?: string | null;
  planned_end_date?: string | null;
  actual_start_date?: string | null;
  actual_end_date?: string | null;
  progress_percent?: number | string;
}

export interface DeliveryProjectNode {
  id: number;
  project_id: number;
  node_key: string;
  node_name: string;
  sort_order: number;
  status: string;
  progress_percent: number | string;
  owner_id?: number | null;
  owner_name?: string | null;
  planned_start_date?: string | null;
  planned_end_date?: string | null;
  actual_start_date?: string | null;
  actual_end_date?: string | null;
  is_critical: boolean;
  is_milestone: boolean;
  tasks?: DeliveryProjectNodeTask[];
}

export interface DeliveryProject {
  id: number;
  project_code: string;
  project_name: string;
  process_template_id?: number | null;
  process_template_name?: string | null;
  sales_order_id?: number | null;
  sales_order_code?: string | null;
  customer_id?: number | null;
  customer_name?: string | null;
  delivery_date?: string | null;
  owner_id?: number | null;
  owner_name?: string | null;
  members?: DeliveryMember[];
  material_code?: string | null;
  material_name?: string | null;
  material_spec?: string | null;
  status: string;
  progress_percent: number | string;
  current_node_key?: string | null;
  current_node_name?: string | null;
  planned_start_date?: string | null;
  planned_end_date?: string | null;
  actual_start_date?: string | null;
  actual_end_date?: string | null;
  notes?: string | null;
  rd_project_id?: number | null;
  nodes?: DeliveryProjectNode[];
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface DeliveryLinkedRdProject {
  id: number;
  project_code: string;
  project_name: string;
}

export interface DeliveryProjectWorkbench extends DeliveryProject {
  recent_reports?: DeliveryNodeReport[];
  open_issues?: DeliveryIssue[];
  linked_rd_project?: DeliveryLinkedRdProject | null;
}

export interface DeliveryProcessTemplateNodeTask {
  id?: number;
  template_node_id?: number;
  task_key: string;
  task_name: string;
  sort_order: number;
  default_owner_role?: string | null;
  planned_duration_days?: number;
}

export interface DeliveryProcessTemplateNode {
  id?: number;
  template_id?: number;
  node_key: string;
  node_name: string;
  sort_order: number;
  default_owner_role?: string | null;
  planned_duration_days: number;
  is_critical: boolean;
  is_milestone: boolean;
  tasks?: DeliveryProcessTemplateNodeTask[];
}

export interface DeliveryProcessTemplate {
  id: number;
  template_code: string;
  template_name: string;
  project_type?: string | null;
  is_active: boolean;
  is_default: boolean;
  notes?: string | null;
  nodes: DeliveryProcessTemplateNode[];
  created_at?: string;
  updated_at?: string;
}

export interface DeliveryNodeReport {
  id: number;
  report_code: string;
  project_id: number;
  project_code: string;
  node_id: number;
  node_key: string;
  node_name: string;
  reporter_name?: string | null;
  report_date: string;
  progress_percent: number | string;
  content?: string | null;
  status: string;
  reviewer_name?: string | null;
  reviewed_at?: string | null;
  review_notes?: string | null;
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface DeliveryIssue {
  id: number;
  issue_code: string;
  project_id: number;
  project_code: string;
  node_id?: number | null;
  node_name?: string | null;
  issue_type: string;
  priority: string;
  status: string;
  title: string;
  description?: string | null;
  assignee_name?: string | null;
  due_date?: string | null;
  resolution?: string | null;
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface DeliveryProgressSummaryRow {
  id: number;
  project_code: string;
  project_name: string;
  customer_name?: string | null;
  sales_order_code?: string | null;
  delivery_date?: string | null;
  owner_name?: string | null;
  material_code?: string | null;
  material_name?: string | null;
  status: string;
  progress_percent: number | string;
  current_node_name?: string | null;
  planned_end_date?: string | null;
  overdue_node_count: number;
  open_issue_count: number;
  days_to_delivery?: number | null;
  node_summary?: string | null;
}

export interface DeliveryGanttItem {
  id: number;
  project_id: number;
  node_id: number;
  project_code: string;
  project_name: string;
  node_name: string;
  customer_name?: string | null;
  node_status?: string | null;
  planned_start_date?: string | null;
  planned_end_date?: string | null;
  progress?: number;
}

export interface DeliveryDashboard {
  kpis: {
    active_projects: number;
    overdue_nodes: number;
    at_risk_projects: number;
    open_issues: number;
  };
  recent_projects: DeliveryProject[];
  overdue_nodes: Array<{
    project_id: number;
    project_code?: string;
    project_name?: string;
    node_id: number;
    node_name: string;
    planned_end_date?: string;
  }>;
  project_gantt?: DeliveryGanttItem[];
}

export interface DeliveryFollowUpRow extends DeliveryProject {
  nodes: DeliveryProjectNode[];
}

export interface DeliveryScheduleRow {
  project_id: number;
  project_code: string;
  project_name: string;
  customer_name?: string | null;
  delivery_date?: string | null;
  owner_name?: string | null;
  status: string;
  progress_percent: number | string;
  current_node_name?: string | null;
  schedule_node_name?: string | null;
  schedule_node_owner_name?: string | null;
  planned_start_date?: string | null;
  planned_end_date?: string | null;
  node_status?: string | null;
  report_overdue: boolean;
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface DeliveryProcessProgressRow {
  id: string;
  project_id: number;
  project_code: string;
  project_name: string;
  sales_order_code?: string | null;
  customer_name?: string | null;
  project_owner_name?: string | null;
  material_name?: string | null;
  delivery_date?: string | null;
  node_id: number;
  node_key: string;
  node_name: string;
  sort_order: number;
  node_status: string;
  progress_percent: number | string;
  node_owner_name?: string | null;
  planned_start_date?: string | null;
  planned_end_date?: string | null;
  actual_start_date?: string | null;
  actual_end_date?: string | null;
  reporter_name?: string | null;
  issue_count: number;
  is_critical: boolean;
  is_milestone: boolean;
}

export interface DeliveryIssueProgressRow {
  id: number;
  issue_code: string;
  project_code: string;
  project_name: string;
  customer_name?: string | null;
  node_name?: string | null;
  issue_type: string;
  priority: string;
  status: string;
  title: string;
  assignee_name?: string | null;
  due_date?: string | null;
  created_at?: string | null;
}

const BASE = '/apps/kuaizhizao';

export const deliveryProjectApi = {
  list: (params?: Record<string, unknown>) =>
    apiRequest<{ items: DeliveryProject[]; total: number }>(`${BASE}/delivery-projects`, { method: 'GET', params }),
  get: (id: number) => apiRequest<DeliveryProject>(`${BASE}/delivery-projects/${id}`, { method: 'GET' }),
  getWorkbench: (id: number) =>
    apiRequest<DeliveryProjectWorkbench>(`${BASE}/delivery-projects/${id}/workbench`, { method: 'GET' }),
  create: (data: Record<string, unknown>) =>
    apiRequest<DeliveryProject>(`${BASE}/delivery-projects`, { method: 'POST', data }),
  update: (id: number, data: Record<string, unknown>) =>
    apiRequest<DeliveryProject>(`${BASE}/delivery-projects/${id}`, { method: 'PUT', data }),
  delete: (id: number) => apiRequest(`${BASE}/delivery-projects/${id}`, { method: 'DELETE' }),
  start: (id: number) => apiRequest<DeliveryProject>(`${BASE}/delivery-projects/${id}/start`, { method: 'POST' }),
  pause: (id: number) => apiRequest<DeliveryProject>(`${BASE}/delivery-projects/${id}/pause`, { method: 'POST' }),
  resume: (id: number) => apiRequest<DeliveryProject>(`${BASE}/delivery-projects/${id}/resume`, { method: 'POST' }),
  cancel: (id: number) => apiRequest<DeliveryProject>(`${BASE}/delivery-projects/${id}/cancel`, { method: 'POST' }),
  complete: (id: number, data?: { force?: boolean; reason?: string }) =>
    apiRequest<DeliveryProject>(`${BASE}/delivery-projects/${id}/complete`, { method: 'POST', data: data ?? {} }),
  changeTemplate: (id: number, processTemplateId: number) =>
    apiRequest<DeliveryProject>(`${BASE}/delivery-projects/${id}/change-template`, {
      method: 'POST',
      data: { process_template_id: processTemplateId },
    }),
  updateNode: (projectId: number, nodeId: number, data: Record<string, unknown>) =>
    apiRequest<DeliveryProjectNode>(`${BASE}/delivery-projects/${projectId}/nodes/${nodeId}`, { method: 'PUT', data }),
  createNodeTask: (projectId: number, data: Record<string, unknown>) =>
    apiRequest<DeliveryProjectNodeTask>(`${BASE}/delivery-projects/${projectId}/node-tasks`, {
      method: 'POST',
      data,
    }),
  updateNodeTask: (projectId: number, taskId: number, data: Record<string, unknown>) =>
    apiRequest<DeliveryProjectNodeTask>(`${BASE}/delivery-projects/${projectId}/node-tasks/${taskId}`, {
      method: 'PUT',
      data,
    }),
  deleteNodeTask: (projectId: number, taskId: number) =>
    apiRequest(`${BASE}/delivery-projects/${projectId}/node-tasks/${taskId}`, { method: 'DELETE' }),
  progressSummary: (params?: Record<string, unknown>) =>
    apiRequest<{ items: DeliveryProgressSummaryRow[]; total: number }>(
      `${BASE}/delivery-reports/progress-summary`,
      { method: 'GET', params },
    ),
  dashboard: () => apiRequest<DeliveryDashboard>(`${BASE}/delivery-dashboard`, { method: 'GET' }),
  followUp: (params?: Record<string, unknown>) =>
    apiRequest<{ items: DeliveryFollowUpRow[]; total: number }>(`${BASE}/delivery-follow-up`, { method: 'GET', params }),
  schedules: (params?: Record<string, unknown>) =>
    apiRequest<{ items: DeliveryScheduleRow[]; total: number }>(`${BASE}/delivery-schedules`, { method: 'GET', params }),
  processProgress: (params?: Record<string, unknown>) =>
    apiRequest<{ items: DeliveryProcessProgressRow[]; total: number }>(
      `${BASE}/delivery-reports/process-progress`,
      { method: 'GET', params },
    ),
  issueProgress: (params?: Record<string, unknown>) =>
    apiRequest<{ items: DeliveryIssueProgressRow[]; total: number }>(
      `${BASE}/delivery-reports/issue-progress`,
      { method: 'GET', params },
    ),
  previewPushFromSalesOrder: (salesOrderId: number) =>
    apiRequest(`${BASE}/sales-orders/${salesOrderId}/push-to-delivery-project/preview`, { method: 'GET' }),
  pushFromSalesOrder: (salesOrderId: number, data?: Record<string, unknown>) =>
    apiRequest<DeliveryProject>(`${BASE}/sales-orders/${salesOrderId}/push-to-delivery-project`, { method: 'POST', data }),
};

export const deliveryProcessTemplateApi = {
  list: (params?: Record<string, unknown>) =>
    apiRequest<{ items: DeliveryProcessTemplate[]; total: number }>(`${BASE}/delivery-process-templates`, { method: 'GET', params }),
  get: (id: number) => apiRequest<DeliveryProcessTemplate>(`${BASE}/delivery-process-templates/${id}`, { method: 'GET' }),
  create: (data: Record<string, unknown>) =>
    apiRequest<DeliveryProcessTemplate>(`${BASE}/delivery-process-templates`, { method: 'POST', data }),
  update: (id: number, data: Record<string, unknown>) =>
    apiRequest<DeliveryProcessTemplate>(`${BASE}/delivery-process-templates/${id}`, { method: 'PUT', data }),
  delete: (id: number) => apiRequest(`${BASE}/delivery-process-templates/${id}`, { method: 'DELETE' }),
  setDefault: (id: number) =>
    apiRequest<DeliveryProcessTemplate>(`${BASE}/delivery-process-templates/${id}/set-default`, { method: 'POST' }),
};

export const deliveryNodeReportApi = {
  list: (params?: Record<string, unknown>) =>
    apiRequest<{ items: DeliveryNodeReport[]; total: number }>(`${BASE}/delivery-node-reports`, { method: 'GET', params }),
  get: (id: number) => apiRequest<DeliveryNodeReport>(`${BASE}/delivery-node-reports/${id}`, { method: 'GET' }),
  create: (data: Record<string, unknown>) =>
    apiRequest<DeliveryNodeReport>(`${BASE}/delivery-node-reports`, { method: 'POST', data }),
  update: (id: number, data: Record<string, unknown>) =>
    apiRequest<DeliveryNodeReport>(`${BASE}/delivery-node-reports/${id}`, { method: 'PUT', data }),
  delete: (id: number) => apiRequest(`${BASE}/delivery-node-reports/${id}`, { method: 'DELETE' }),
  submit: (id: number) => apiRequest<DeliveryNodeReport>(`${BASE}/delivery-node-reports/${id}/submit`, { method: 'POST' }),
  review: (id: number, data: { approved: boolean; review_notes?: string }) =>
    apiRequest<DeliveryNodeReport>(`${BASE}/delivery-node-reports/${id}/review`, { method: 'POST', data }),
};

export const deliveryIssueApi = {
  list: (params?: Record<string, unknown>) =>
    apiRequest<{ items: DeliveryIssue[]; total: number }>(`${BASE}/delivery-issues`, { method: 'GET', params }),
  get: (id: number) => apiRequest<DeliveryIssue>(`${BASE}/delivery-issues/${id}`, { method: 'GET' }),
  create: (data: Record<string, unknown>) =>
    apiRequest<DeliveryIssue>(`${BASE}/delivery-issues`, { method: 'POST', data }),
  update: (id: number, data: Record<string, unknown>) =>
    apiRequest<DeliveryIssue>(`${BASE}/delivery-issues/${id}`, { method: 'PUT', data }),
  delete: (id: number) => apiRequest(`${BASE}/delivery-issues/${id}`, { method: 'DELETE' }),
};

export const DELIVERY_PROJECT_STATUS: Record<string, string> = {
  draft: '草稿',
  in_progress: '进行中',
  paused: '暂停',
  completed: '已完成',
  cancelled: '已取消',
};

export const DELIVERY_NODE_STATUS: Record<string, string> = {
  not_started: '未开始',
  in_progress: '进行中',
  completed: '已完成',
  overdue: '逾期',
};

export const DELIVERY_NODE_TASK_STATUS: Record<string, string> = {
  todo: '待办',
  in_progress: '进行中',
  done: '已完成',
  cancelled: '已取消',
};

export const DELIVERY_NODE_REPORT_STATUS: Record<string, string> = {
  draft: '草稿',
  submitted: '已提交',
  approved: '已通过',
  rejected: '已驳回',
};

export const DELIVERY_ISSUE_STATUS: Record<string, string> = {
  open: '待处理',
  in_progress: '处理中',
  resolved: '已解决',
  closed: '已关闭',
};

export const DELIVERY_ISSUE_PRIORITY: Record<string, string> = {
  low: '低',
  normal: '普通',
  high: '高',
  urgent: '紧急',
};

export const DELIVERY_ISSUE_TYPE: Record<string, string> = {
  blocker: '阻塞',
  quality: '质量',
  delivery: '交期',
  other: '其他',
};
